specs/Standards/Contracts/SelfIdentification.md

```diff
@@ -0,0 +1,208 @@
+
+To satisfy this NEP the contract must implement 1 (one) method: 
+
+1. `get_contract_info() -> ContractInfo` returning a JSON structure as described below (typescript):
+
+```typescript
+type ContractInfo = {
+   dataVersion: number = 1; // determines the rest of the fields
+   webAppUrl: string = ""; // url for the webApp/DApp for this contract
+   name: string = ""; // contract-code short name
+   version: string = "0.0.1"; //contract-code semver
+   source: string = ""; //contract source code URL, e.g. http://github.com/author/contract
+   webAppUrl: string=""; //web URL where the DApp can be accessed e.g. http://Narwallets.com/div-pool/
+   standards: string[] = []; // standards this contract satisfies
+   developersAccountId: string=""; //near account of the author of the code
+   auditorAccountId: string=""; //near account of the auditor of the code
+}
+```
+
+### Details on the fields:
+- **dataVersion**. In future iterations of this NEP, new fields migth be added to `ContractInfo`, this number will be increased. All `dataVersions` are backwards compatible, meaning `dataVersion=2` **can only add fields** to `dataVersion=1`'s structure
+
+### Use Cases 
+
+Alice wants to trade only with audited NEP-21 tokens. She receives an account name as a token source. Before interacting with the account her code can:
+
+* perform basic account validation (check that there are no full-access keys on the contract)
+
+* call `get_contract_info`
+
+* verify the contract supports NEP-21 `standards.include("NEP-21")`
+
+* use `auditorAccountId` to check the contracts code hash and audit info (depending on the auditor's mechanism to do that)
+
+If all the above is correct, her code adds the contract as an audited NEP-21 Token and the user can interact with it via `contractInfo.webAppUrl`
+
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Reference Implementations:
+
+#### Rust:
+```rust
+// ---------------------------------
+// -- CONTRACT Self Identification --
+// ---------------------------------
+// [NEP-129](https://github.com/nearprotocol/NEPs/pull/129)
+// see also pub fn get_contract_info
+pub const CONTRACT_NAME: &str = "diversifying staking pool";
+pub const CONTRACT_VERSION: &str = "0.1.0";
+pub const DEVELOPERS_ACCOUNT_ID: &str = "developers.near"; 
+pub const DEFAULT_WEB_APP_URL: &str = "http://div-pool.narwallets.com";
+pub const DEFAULT_AUDITOR_ACCOUNT_ID: &str = "auditors.near";
+
+/// NEP-129 get information about this contract
+/// returns JSON string according to [NEP-129](https://github.com/nearprotocol/NEPs/pull/129)
+/// Rewards fee fraction structure for the staking pool contract.
+#[derive(Serialize)]
+#[serde(crate = "near_sdk::serde")]
+#[allow(non_snake_case)]
+pub struct NEP129Response {
+    pub dataVersion:u16,
+    pub name:String,
+    pub version:String,
+    pub developersAccountId:String,
+    pub source:String,
+    pub standards:Vec<String>,
+    pub webAppUrl:String,
```

Comment:
`webAppUrl` should be optional (`Option<String>`) because not all contracts will have a web app frontend. Optional fields should be always be explicit in the code.

Reply:
Ok. It applies also for auditor_account_id

Reply:
done. also for auditor_account_id

Reply:
I would actually suggest to maintain audit map somewhere outside of the contract itself.
There can be multiple audits for the same contract, audits should contain some extra info (like advisories) and audits can frequently come in after the contract is deployed. Also it's ideally that audit itself shouldn't modify the commit hash so can rely on it, which this would potentially lead to.

We actually have an audit registry built (but not launched on MainNet yet) - explained more here - https://gov.near.org/t/audit-registry-add-accountability-and-visibility-of-which-applications-were-reviewed-and-by-whom/234